engine/access/ingestion2/results_forest.go

```diff
@@ -0,0 +1,452 @@
+// for a particular ExecutionResult. The mempool is aware of the tree structure
+// formed by the results. The mempool supports pruning by view: only results
+// descending from the latest persisted sealed result are relevant. Hence, we
+// can prune all results for blocks _below_ the latest persisted sealed result.
+// Results of sufficient views for forks that conflict with the finalized fork are
+// retained. However, such orphaned forks do not grow anymore and their results
+// will be progressively flushed out with increasing sealed view.
+//
+// Safe for concurrent access. Internally, the mempool utilizes the LevelledForrest.
+// For an in-depth discussion of the core algorithm, see ./Fork-Aware_Mempools.md
+type ResultsForest struct {
+	component.Component
+
+	log zerolog.Logger
+	cm  *component.ComponentManager
+
+	forest forest.LevelledForest
+
+	// maxRunningCount is the maximum number of pipelines that can be run in parallel.
+	maxRunningCount uint
+
+	// latestSealedResultID is the ID of vertex that was most recently marked sealed.
+	// This may be the same as the latest persisted sealed result ID, or a descendent.
+	// This is used by the sealing logic to backfill any results that were sealed during loading.
+	latestSealedResultID flow.Identifier
+
+	// latestPersistedSealedResult holds the resultID of the latest sealed result that is fully indexed
+	// and persisted to the database.
+	latestPersistedSealedResult *LatestPersistedSealedResult
+
+	// loadingComplete is set to true when all results have been loaded from the database into the forest.
+	loadingComplete bool
+
+	// notifier is used to signal to the pipelineManager>Loop that a new vertex is available, or one has completed
+	notifier engine.Notifier
+
+	mu sync.RWMutex
+}
+
+// NewResultsForest instantiates a ResultsForest
+func NewResultsForest(
+	log zerolog.Logger,
+	latestPersistedSealedResult *LatestPersistedSealedResult,
+) *ResultsForest {
+	f := &ResultsForest{
+		log:                         log.With().Str("component", "results_forest").Logger(),
+		forest:                      *forest.NewLevelledForest(0),
+		maxRunningCount:             DefaultMaxProcessingSize,
+		latestSealedResultID:        latestPersistedSealedResult.ResultID(),
+		latestPersistedSealedResult: latestPersistedSealedResult,
+		notifier:                    engine.NewNotifier(),
+	}
+
+	cm := component.NewComponentManagerBuilder().
+		AddWorker(f.pipelineManagerLoop).
+		Build()
+
+	f.Component = cm
+	f.cm = cm
+
+	return f
+}
+
+func (f *ResultsForest) pipelineManagerLoop(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	wg := sync.WaitGroup{}
+	ch := f.notifier.Channel()
+	for {
+		select {
+		case <-ctx.Done():
+			// wait for all pipelines to gracefully shutdown
+			wg.Wait()
+			return
+
+		case <-ch:
+			runningCount := uint(0)
+
+			// Inspect all vertices in the tree forming from the latest persisted sealed result.
```

Comment:
I am wondering what you think about the alternative approach: 
* overall, I would prefer if that thread and work management lived in the [Engine](https://github.com/onflow/flow-go/blob/ec740f7ebb96f4f4eb92796b7f305e8462af5f36/engine/access/ingestion2/engine.go#L20)
*  Assume that when we call [`ResultsForest.AddResult`](https://github.com/onflow/flow-go/blob/391632a0525f8afd694d3aca312a3ae18586725a/engine/access/ingestion2/results_forest.go#L197) we create a new task and add it to some queue. Then we have a worker pool with 20 workers pulling elements of that queue.

Reply:
I'll look into this. I'd have to use a priority queue based on height, but that should be simple enough.